Client/Frontend/Settings/SettingsTableViewController.swift

```diff
@@ -194,6 +194,135 @@ class BoolSetting: Setting {
     }
 }
 
+// A helper class for a setting backed by a UITextField.
+// This takes an optional settingIsValid and settingDidChange callback
+// If settingIsValid returns false, the Setting will not change and the text remains red.
+class StringSetting: Setting, UITextFieldDelegate {
+
+    let prefKey: String
+
+    private let prefs: Prefs
+    private let defaultValue: String?
+    private let placeholder: String?
+    private let settingDidChange: ((String?) -> Void)?
+    private let settingIsValid: ((String?) -> Bool)?
```

Comment:
Nits: You can reduce some parenthesage (yes I made that word up) by removing parens around single-argument closures (e.g., `(String? -> Void)?`).